content/blog/typescript-type-generics/index.md

```diff
@@ -0,0 +1,340 @@
+
+For example, if we wanted to pass in an object of some kind (such as `{}`, a simple empty object), it would be invalid:
+
+```typescript
+returnProp({}) // Argument of type '{}' is not assignable to parameter of type 'string'.
+```
+
+This may seem obvious from the typings, but _we ideally want `returnProp` to accept ANY type because **we aren't using any operations that require knowing the type**._ (no addition or subtraction, requiring a number; no string concatenation that might restrict an object from being passed).
+
+## Potential Solution 3: Any {#generic-usecase-setup-any-solution}
+
+Of course, we could use the `any` in order to force any input and return type. (Goodness knows I've had my fair share of typing frustrations that ended with a few `any`s in my codebase!)
+
+Although this would allow any input type, we'd also be losing any type information between the input and output. As a result our types would be too loose on the return type:
+
+```typescript
+function returnSelf(returnProp: any): any {
+    return returnProp;
+}
+
+const returnedObject = returnSelf({objProperty: 12}); // This now works! 🎉
+
+returnedObject.test(); // This will not return an error but should 🙁
+returnedObject.objProperty; // This will also (correctly) not throw an error, but TS won't know it's a number ☹️
+```
+
+# The Real Solution {#generics-intro}
+
+So what's the answer? How can we get preserved type data on both the input and the output??
+
+The solution is... Well, you've read the title I'm sure.
+
+_Type generics allow us to store loose type data in a **type variable**_. A type variable is _a unique kind of variable that's not exposed to JavaScript but is instead handled by TypeScript to provide expected typing data_. For example, the above example could be rewritten as:
+
+```typescript
+function returnSelf<T>(returnProp: T): T {
+    return returnProp;
+}
+```
+
+In this example, we're defining a type variable `T`, then telling TS that both the property and the return type should be the same type.
+
+This means that you can use the function like this:
+
+```typescript
+const numberVar = returnSelf(2); // T in this instance is `2`, so it's similar to writing `const numberVal: 2 = 2;`
+
+// Likewise, this object is now returned as if it was just placed on the const
+const returnedObject = returnSelf({objProperty: 12});
+
+// This will fail, as expected
+returnedObject.test();
+// This will exist, and TS will know it as a number
+returnedObject.objProperty;
+```
+
+> Author's note:
+>
+> The type variable does not need to be called `T`. In fact, while it seems to be commonplace for the community to use single-letter type variable names (often due to the length and complexity of the typings), there are many reasons why more explicit type names should be used.
+>
+> Remember, type variables are like other variables in that you need to maintain them and understand what they're doing in your code.
```

Comment:
I love this. Thanks for adding this touch. 🙏🙏🙏

Reply:
I wanted to make sure to bring it up since I think it's important. I want to make sure in more complex articles to move away from  `T` as well, but I think in an intro to generics seeing `T` might help understand conventions and keep an idea of abstract type concepts